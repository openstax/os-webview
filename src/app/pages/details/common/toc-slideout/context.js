import buildContext from '~/components/jsx-helpers/build-context';
import {useToggle} from '~/helpers/data';

function useContextValue() {
    const [tocActive, toggleTocActive] = useToggle(false);

    return {
        isOpen: tocActive,
        toggle: toggleTocActive
    };
}

const {useContext, ContextProvider} = buildContext({useContextValue});

export {
    useContext as default,
    ContextProvider as TOCContextProvider
};
