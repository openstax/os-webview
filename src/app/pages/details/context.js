import React from 'react';
import buildContextLoader from '~/components/jsx-helpers/context-loader';
import buildContext from '~/components/jsx-helpers/build-context';

/*
    Managing the slug requires a Context
*/
function getSlugFromLocation() {
    const bookTitle = window.location.pathname.replace(/.*details\//, '');
    let slug;

    if ((/^books/).test(bookTitle)) {
        slug = bookTitle;
    } else {
        slug = `books/${bookTitle}`;
    }
    // Special handling for books whose slugs have changed
    if ((/university-physics$/).test(slug)) {
        slug += '-volume-1';
    }

    return slug;
}

function useSlugContextValue() {
    const [slug, setSlug] = React.useState(getSlugFromLocation());

    React.useEffect(() => {
        const updateSlug = () => {
            setSlug(getSlugFromLocation());
        };

        window.addEventListener('popstate', updateSlug);

        return () => window.removeEventListener('popstate', updateSlug);
    });

    return {slug, setSlug};
}

const {
    useContext: useSlugContext,
    ContextProvider: SlugContextProvider
} = buildContext({useContextValue: useSlugContextValue});

// ------

function useContextValue(data) {
    data.comingSoon = data.bookState === 'coming_soon';
    data.language = data.meta.locale;

    return data;
}

const {useContext, ContextLoader} = buildContextLoader();

function DetailsLoader({children}) {
    const {slug} = useSlugContext();

    return (
        <ContextLoader
            slug={slug}
            useContextValue={useContextValue}
            doDocumentSetup
        >
            {children}
        </ContextLoader>
    );
}

export function DetailsContextProvider({children}) {
    return (
        <SlugContextProvider>
            <DetailsLoader>
                {children}
            </DetailsLoader>
        </SlugContextProvider>
    );
}

export default useContext;
