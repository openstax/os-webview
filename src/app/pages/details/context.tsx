import {useState} from 'react';
import buildContext from '~/components/jsx-helpers/build-context';
import type {ResourceData} from './common/resource-box/resource-box-utils';
import type {PromoteData} from './desktop-view/promo';

export type LocaleType = {
    locale: string;
};
export type TranslationType = Array<LocaleType & {slug: string}>;
type WebinarContent = {
    content: {
        heading: string;
        content: string;
    };
    link: string;
};
type StuffContent = {
    content: {
        id: number;
        buttonText: string | null;
        buttonUrl: string | null;
        content: string;
        contentLoggedIn: string | null;
        heading: string;
    };
};
type VideoContent = {
    title: string;
    description: string;
    embed: string;
};
type Author = {
    name: string;
    university: string;
    seniorAuthor?: boolean;
};

export type IsbnType = 'print' | 'printSoftcover' | 'digital' | 'ibook' | 'ibookVolume2';

export type ContextValues = {
    id: number;
    slug: string;
    translations: Array<TranslationType>;
    bookState: string;
    comingSoon: boolean;
    coverColor: string;
    meta: LocaleType;
    reverseGradient: boolean;
    title: string;
    titleImageUrl: string;
    communityResourceHeading?: string;
    communityResourceLogoUrl?: string;
    communityResourceUrl?: string;
    communityResourceCta?: string;
    communityResourceBlurb: string;
    communityResourceFeatureLinkUrl?: string;
    communityResourceFeatureText?: string;
    amazonIframe: string | null;
    webinarContent?: WebinarContent;
    freeStuffStudent: StuffContent;
    freeStuffInstructor: StuffContent;
    videos: VideoContent[];
    setUseCardBackground?: React.Dispatch<React.SetStateAction<boolean>>;
    authors: Author[];
    created: string;
    updated: string;
    coverUrl: string;
    webviewRexLink: string;
    webviewLink: string;
    errataContent?: string;
    cnxId: string;
    publishDate: string;
    printIsbn10?: string;
    printIsbn13: string;
    printSoftcoverIsbn10?: string;
    printSoftcoverIsbn13: string;
    digitalIsbn10?: string;
    digitalIsbn13: string;
    ibookIsbn10?: string;
    ibookIsbn13: string | null;
    ibookVolume2Isbn10?: string;
    ibookVolume2Isbn13: string | null;
    lastUpdatedWeb?: string;
    lastUpdatedPdf: string | null;
    licenseName: string;
    licenseText: string;
    licenseTitle?: string;
    licenseVersion: string;
    licenseIcon?: string;
    bookStudentResources: ResourceData[];
    promoteSnippet: PromoteData[];
    featuredResourcesHeader: string | null;
    partnerListLabel: string;
    partnerPageLinkText: string;
    salesforceAbbreviation: string;
    adoptions?: number;
    supportStatement: string;
    savings: number;
};

function useContextValue({data}: {data: ContextValues}) {
    const [useCardBackground, setUseCardBackground] = useState(false);

    data.comingSoon = data.bookState === 'coming_soon';

    return {...data, useCardBackground, setUseCardBackground};
}

const {useContext, ContextProvider} = buildContext({useContextValue});

export {useContext as default, ContextProvider as DetailsContextProvider};
