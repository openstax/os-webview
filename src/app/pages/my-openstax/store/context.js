import React from 'react';
import buildContext from '~/components/jsx-helpers/build-context';
import sfApiFetch from '~/models/sfapi';
import {camelCaseKeys} from '~/helpers/page-data-utils';
import uniq from 'lodash/uniq';

// Takes a school entry from sfapi, returns a standard school
function sfSchoolToInstitution(school) {
    return ({
        id: school.salesforce_id,
        name: school.name || '[unknown school]'
    });
}

function useContextValue() {
    const [lists, setLists] = React.useState([]);
    const [schools, setSchools] = React.useState({});
    const decode = React.useCallback(
        (idList) => {
            const toLookUp = uniq(idList)
                .filter((id) => typeof id !== 'undefined' && !(id in schools));

            if (toLookUp.length === 0) {
                return;
            }

            // Put placeholders in the state
            const placeholders = toLookUp.reduce(
                (a, id) => {
                    a[id] = '';
                    return a;
                },
                {}
            );

            setSchools({...schools, ...placeholders});

            // Look them up async
            const promises = toLookUp.map((id) => sfApiFetch('schools', `/${id}`));

            Promise.all(promises).then((results) => {
                const newEntries = results
                    .map(sfSchoolToInstitution)
                    .reduce((a, b) => {
                        a[b.id] = b;
                        return a;
                    }, {});

                setSchools({...schools, ...newEntries});
            });
        },
        [schools]
    );

    React.useEffect(
        () => {
            sfApiFetch('lists')
                .then(camelCaseKeys)
                .then(setLists);
        },
        []
    );

    return {lists, schools, decode};
}

const {useContext, ContextProvider} = buildContext({useContextValue});

export {
    useContext as default,
    ContextProvider as MyOpenStaxContextProvider
};
