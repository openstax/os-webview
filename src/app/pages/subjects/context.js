import React from 'react';
import buildContext from '~/components/jsx-helpers/build-context';
import usePageData from '~/components/jsx-helpers/page-loader';
import useLanguageContext from '~/models/language-context';

const preserveWrapping = true;

function useContextValue() {
    const [slug, setSlug] = React.useState('subjects');
    const data = usePageData(`pages/${slug}`, preserveWrapping);
    const {language} = useLanguageContext();

    React.useEffect(() => {
        if (!data || !data.translations) {
            return;
        }
        const translations = data.translations[0].value;
        const thisTranslation = translations.find((t) => t.locale === language);

        if (thisTranslation) {
            setSlug(thisTranslation.slug);
        }
    }, [data, language]);

    if (data) {
        data.books = data.books.filter((b) => b.bookState !== 'retired');
    }

    return data;
}

const {useContext, ContextProvider} = buildContext({useContextValue});

export {
    useContext as default,
    ContextProvider as SubjectsContextProvider
};
