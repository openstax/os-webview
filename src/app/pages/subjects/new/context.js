import React from 'react';
import buildContext from '~/components/jsx-helpers/build-context';
import usePageData from '~/components/jsx-helpers/page-loader';
import useLanguageContext from '~/contexts/language';

const preserveWrapping = true;
const icons = [
    '/dist/images/subjects/author-icon.svg',
    '/dist/images/subjects/scope-icon.svg',
    '/dist/images/subjects/review-icon.svg'
];

function aboutBlurbs(model) {
    const textData = Reflect.ownKeys(model)
        .filter((k) => (/^devStandard\d/).test(k))
        .reduce((a, b) => {
            const [_, num, textId] = b.match(/(\d+)(\w+)/);
            const index = num - 1;

            a[index] = a[index] || {};
            a[index][textId.toLowerCase()] = model[b];
            a[index].iconUrl = icons[index];
            return a;
        }, []);

    return textData;
}

function useContextValue() {
    const [slug, setSlug] = React.useState('new-subjects');
    const data = usePageData(`pages/${slug}`, preserveWrapping);
    const {language} = useLanguageContext();

    React.useEffect(() => {
        if (!data || !data.translations || !data.translations.length) {
            return;
        }
        const translations = data.translations[0].value;
        const thisTranslation = translations.find((t) => t.locale === language);

        if (thisTranslation) {
            setSlug(thisTranslation.slug);
        }
    }, [data, language]);

    if (data?.books) {
        data.books = data.books.filter((b) => b.bookState !== 'retired');
    }

    if (data) {
        data.aboutBlurbs = aboutBlurbs(data);
    }

    return data;
}

const {useContext, ContextProvider} = buildContext({useContextValue});

export {
    useContext as default,
    ContextProvider as SubjectsContextProvider
};
