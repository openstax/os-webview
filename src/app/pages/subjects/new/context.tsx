import React from 'react';
import buildContext from '~/components/jsx-helpers/build-context';
import usePageData from '~/helpers/use-page-data';
import useLanguageContext from '~/contexts/language';
import type {LocaleEntry} from '~/components/language-selector/language-selector';
import type {InfoBox} from './info-boxes';
import type {TutorValue} from './tutor-ad';
import {toNumber} from 'lodash';

type DevStandardKeys = 'devStandard1Heading';
type DevStandardPair = {[key in DevStandardKeys]: string};
export type SubjectData = {
    icon: string;
    categories: string[];
};
export type ImageData = {
    id: number;
    title: string;
    file: string;
};
type SubjectsPageData = {
    title: string;
    pageDescription: string;
    subjects: {[key: string]: SubjectData};
    translations: Array<{value: LocaleEntry[]}>;
    books?: {
        bookState: string;
    }[];
    aboutBlurbs?: ReturnType<typeof aboutBlurbs>;
    heading: string;
    description: string;
    headingImage: {
        meta: {
            downloadUrl: string;
        };
    };
    infoBoxes: InfoBox[];
    tutorAd: [
        {
            value: TutorValue;
        }
    ];
    philanthropicSupport: string;
};

// The Page data before DevStandardPair is translated to aboutBlurbs
type RawPageData = SubjectsPageData & DevStandardPair;

const preserveWrapping = true;
const icons = [
    '/dist/images/subjects/author-icon.svg',
    '/dist/images/subjects/scope-icon.svg',
    '/dist/images/subjects/review-icon.svg'
];

function aboutBlurbs(model: RawPageData) {
    const textData = (Reflect.ownKeys(model) as string[])
        .filter((k) => (/^devStandard\d/).test(k))
        .reduce((a, b) => {
            const [_, num, textId] = b.match(/(\d+)(\w+)/) as string[];
            const index = toNumber(num) - 1;

            a[index] = a[index] || {};
            a[index][textId.toLowerCase()] = model[b as DevStandardKeys];
            a[index].iconUrl = icons[index];
            return a;
        }, [] as {[key: string]: string}[]);

    return textData;
}

function useContextValue() {
    const [slug, setSlug] = React.useState('new-subjects');
    const data = usePageData<RawPageData>(`pages/${slug}`, preserveWrapping);
    const {language} = useLanguageContext();

    React.useEffect(() => {
        if (!data || !data.translations || !data.translations.length) {
            return;
        }
        const translations = data.translations[0].value;
        const thisTranslation = translations.find((t) => t.locale === language);

        if (thisTranslation) {
            setSlug(thisTranslation.slug);
        }
    }, [data, language]);

    if (data?.books) {
        data.books = data.books.filter((b) => b.bookState !== 'retired');
    }

    if (data) {
        data.aboutBlurbs = aboutBlurbs(data);
    }

    return data as SubjectsPageData;
}

const {useContext, ContextProvider} = buildContext({useContextValue});

export {useContext as default, ContextProvider as SubjectsContextProvider};
