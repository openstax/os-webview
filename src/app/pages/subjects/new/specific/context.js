import {useEffect} from 'react';
import usePageData from '~/components/jsx-helpers/page-loader';
import buildContext from '~/components/jsx-helpers/build-context';
import $ from '~/helpers/$';

const preserveWrapping = false;

function useContextValue(slug) {
    const data = usePageData(`pages/${slug}`, preserveWrapping);

    useEffect(
        () => {
            if (data) {
                $.setPageTitleAndDescriptionFromBookData(data);
            }
        },
        [data]
    );

    return data;
}

const {useContext, ContextProvider} = buildContext({useContextValue});

export {
    useContext as default,
    ContextProvider as SpecificSubjectContextProvider
};
