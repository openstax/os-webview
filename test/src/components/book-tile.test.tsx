import React from 'react';
import {describe, expect, it} from '@jest/globals';
import {render, screen} from '@testing-library/preact';
import ShellContextProvider from '~/../../test/helpers/shell-context';
import {Book as BookInfo} from '~/pages/subjects/new/specific/context';
import BookTile from '~/components/book-tile/book-tile';
import userEvent from '@testing-library/user-event';

const bookData: BookInfo = {
    id: 46,
    slug: 'sample',
    title: 'Sample Title',
    webviewRexLink: 'link-to-rex',
    webviewLink: 'osweb-link',
    highResolutionPdfUrl: 'hi-res.pdf',
    lowResolutionPdfUrl: '',
    coverUrl: 'cover.jpg'
};

function Component({book}: {book: [BookInfo]}) {
    return (
        <ShellContextProvider>
            <BookTile book={book} />
        </ShellContextProvider>
    );
}
jest.mock('~/helpers/page-data-utils', () => ({
    ...jest.requireActual('~/helpers/page-data-utils'),
    useDataFromSlug: jest.fn()
}));

/* eslint-disable camelcase */
jest.mock('~/models/book-titles', () => ({
    ...jest.requireActual('~/models/book-titles'),
    __esModule: true,
    default: Promise.resolve([
        {
            id: 46,
            title: 'Prealgebra',
            book_state: 'live',
            promote_snippet: [
                {
                    type: 'content',
                    value: {
                        id: 1,
                        name: 'Assignable',
                        image: 'https://assets.openstax.org/oscms-dev/media/original_images/icons8-image-64.png'
                    },
                    id: 'd6d635c4-990a-44dd-a76a-d007255ba293'
                },
                {
                    type: 'content',
                    value: {
                        id: 2,
                        name: 'Staxly',
                        image: null
                    },
                    id: 'c7c72949-d03f-41fa-93ff-2a5270bba13c'
                }
            ]
        },
        {
            id: 130,
            title: 'Elementary Algebra',
            book_state: 'coming_soon',
            promote_snippet: []
        }
    ])
}));

describe('book-tile', () => {
    it('renders', () => {
        const {baseElement} = render(<Component book={[bookData]} />);

        expect(baseElement).toMatchSnapshot();
    });
    it('renders coming soon', async () => {
        render(<Component book={[{...bookData, id: 130}]} />);

        await screen.findByText('Coming soon');
    });
    it('falls back on webviewLink', async () => {
        const save = bookData.webviewRexLink;

        bookData.webviewRexLink = '';
        render(<Component book={[bookData]} />);
        const user = userEvent.setup();

        await user.click(screen.getByRole('button'));

        const links = screen.getAllByRole('link');

        expect(links).toHaveLength(1);

        await user.click(screen.getAllByRole('link')[0]);

        bookData.webviewRexLink = save;
    });
    it('falls back on lowResolutionPdfUrl (may be obsolete)', async () => {
        const save = bookData.highResolutionPdfUrl;

        bookData.lowResolutionPdfUrl = 'low-res.pdf';
        bookData.highResolutionPdfUrl = '';

        render(<Component book={[bookData]} />);

        const user = userEvent.setup();

        const links = screen.getAllByRole('link');

        expect(links).toHaveLength(1);

        await user.click(screen.getAllByRole('link')[0]);

        bookData.highResolutionPdfUrl = save;
    });
});
