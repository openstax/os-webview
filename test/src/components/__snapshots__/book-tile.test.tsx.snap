// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`book-tile renders 1`] = `
<body>
  <div>
    <div
      class="book-tile"
    >
      <a
        aria-label="Sample Title book"
        href="/details/sample"
      >
        <img
          height="240"
          role="presentation"
          src="cover.jpg"
          width="240"
        />
        <div
          class="text-block"
        >
          Sample Title
        </div>
      </a>
      <div
        class="navmenu"
        data-analytics-nav="Get the book dropdown"
      >
        <button
          aria-controls="sample-ddm"
          aria-expanded="false"
          aria-haspopup="true"
          id="sample-ddb"
          type="button"
        >
          Get the book
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-caret-down "
            data-icon="caret-down"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 320 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M137.4 374.6c12.5 12.5 32.8 12.5 45.3 0l128-128c9.2-9.2 11.9-22.9 6.9-34.9s-16.6-19.8-29.6-19.8L32 192c-12.9 0-24.6 7.8-29.6 19.8s-2.2 25.7 6.9 34.9l128 128z"
              fill="currentColor"
            />
          </svg>
        </button>
        <div
          aria-labelledby="sample-ddb"
          id="sample-ddm"
          role="menu"
        >
          <a
            href="link-to-rex"
            role="menuitem"
          >
            View online
          </a>
          <a
            href="hi-res.pdf"
            role="menuitem"
          >
             Download a PDF
          </a>
          <hr />
          <a
            href="/details/sample?Instructor resources"
            role="menuitem"
          >
            Instructor resources
          </a>
          <a
            href="/details/sample?Student resources"
            role="menuitem"
          >
            Student resources
          </a>
        </div>
      </div>
    </div>
  </div>
  <div
    class="ReactModalPortal"
  />
  <div
    class="ReactModalPortal"
  />
</body>
`;
